quinn-proto/src/connection/mod.rs

```diff
@@ -673,28 +673,49 @@ where
 
             if close {
                 trace!("sending CONNECTION_CLOSE");
-                match self.state {
-                    State::Closed(state::Closed { ref reason }) => {
-                        if space_id == SpaceId::Data {
-                            reason.encode(&mut buf, builder.max_size)
-                        } else {
-                            frame::ConnectionClose {
-                                error_code: TransportErrorCode::APPLICATION_ERROR,
-                                frame_type: None,
-                                reason: Bytes::new(),
+                // Encode ACKs before the ConnectionClose message, to give the receiver
+                // a better approximate on what data has been processed. This is
+                // especially important with ack delay, since the peer might not
+                // have gotten any other ACK for the data earlier on.
+                if !self.spaces[space_id].pending_acks.ranges().is_empty() {
+                    Self::populate_acks(
+                        self.receiving_ecn,
+                        now,
+                        &mut SentFrames::default(),
+                        &mut self.spaces[space_id],
+                        &mut buf,
+                        &mut self.stats,
+                    );
+                }
+
+                if buf.len() + frame::ConnectionClose::SIZE_BOUND < builder.max_size {
+                    // TODO: Under very rare conditions the ACKs might take up so
+                    // much space that the actual close reason gets swallowed.
+                    // We could send another packet for it, but for this unlikely
+                    // condition we can also just send nothing and let the peer time out.
```

Comment:
I'm not sure this is actually possible; we know we're sending at most 64 ACK ranges, and `ConnectionClose` already truncates the reason if necessary. Given the minimum MTU, I think we can rely on there being enough space for the critical parts.

Reply:
Oh that's true. I initially had nothing, but then felt bad about potential crashes due to having no bounds check.
Probably fine to just replace the comment with "There will always be enough space"?

Reply:
Sure, maybe plus a `debug_assert`.